src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2549,14 +2549,54 @@ bool IsInterfaceImplementationMethodNeededByTypeDueToInterface (OverrideInformat
 		{
 			var @base = overrideInformation.Base;
 			var method = overrideInformation.Override;
+			Debug.Assert (@base.DeclaringType.IsInterface);
 			if (@base is null || method is null || @base.DeclaringType is null)
 				return false;
 
 			if (Annotations.IsMarked (method))
 				return false;
 
-			if (!@base.DeclaringType.IsInterface)
-				return false;
+			// If the override is a DIM that provides an implementation for a type that requires the interface, we may need to mark the DIM
+			var dims = Annotations.GetDefaultInterfaceImplementations (@base).Where(dim => Annotations.IsRelevantToVariantCasting (dim.InstanceType));
+			if (dims.Any (dim => dim.DefaultInterfaceMethod == method && Annotations.IsRelevantToVariantCasting (dim.InstanceType)))
+			{
+				// We need to find the most derived DIM for each type that has DIMs -- if there is a most derived DIM, we only mark that DIM
```

Comment:
This does actually work, the IMiddle.Value in StaticDefaultInterfaceMethodOnDerivedInterface.cs is kept because of the base.IsAbstract check below, but it shouldn't have to. Regardless, getting this to work for instance DIMs might require a larger change and some thought, and a lot more test cases, so I'll separate that out into a new branch and put it in the backlog.